### ISSUE TITLE
Undefined properties cause an exception to be thrown

### ISSUE THREAD

**Issue by davidpolberger** [Starcounter-Jack/JSON-Patch#32]
Consider the following code:

```js
var object = {
  test: undefined
};

var observer = jsonpatch.observe(object);
var result = jsonpatch.generate(observer);
```

JSON-Patch throws a SyntaxError exception when trying to generate the patch at line 421 of json-patch-duplex.js (in the function <code>_generate</code>):

```js
mirror[key] = JSON.parse(JSON.stringify(obj[key]));
```

<code>obj[key]</code> is <code>undefined</code>, causing <code>JSON.stringify()</code> to return <code>undefined</code>, which <code>JSON.parse()</code> cannot handle ("Unexpected token u").

The expected behavior is that no patches should be generated, as nothing has changed.

**Comment by warpech** [comment]
We could handle it somehow in `generate` step. My idea is that when a change to `undefined` is detected, `jsonpatch.generate` reports it as a `remove` operation. What do you think?

**Comment by jmeas** [comment]
@sonnyp [has a convincing argument](https://github.com/Starcounter-Jack/Fast-JSON-Patch/pull/36#issuecomment-46680484) that we should not handle invalid JSON values. I've been persuaded. What do you think, @warpech?

**Comment by sonnyp** [comment]
@jmeas We shouldn't handle undefined in a JSON Patch document (see https://github.com/Starcounter-Jack/Fast-JSON-Patch/issues/32), however not sure what to do about a undefined in a JavaScript object.

**Comment by jmeas** [comment]
Ah, okay. Sounds good :+1:

**Comment by tomalec** [comment]
I've just added a description at [README.md](https://github.com/Starcounter-Jack/JSON-Patch/blob/master/README.md#undefineds-json-to-js-extension), and bunch of tests.

Basically:
 1. For valid JSON document, we will generate only valid JSON Patches,
 2. For invalid JSONs (ones with `undefined`), we will handle it gently. [`.apply`](http://starcounter-jack.github.io/JSON-Patch/test/SpecRunnerDuplex.html?spec=undefined%20-%20JS%20to%20JSON%20projection%20%2F%20JSON%20to%20JS%20extension), `.observe`, `.compare`, [`.generate`](http://starcounter-jack.github.io/JSON-Patch/test/SpecRunnerDuplex.html?spec=duplex%20generate%20undefined%20-%20JSON%20to%20JS%20extension) will use and produce JSON Patches with undefined as with any other falsy value,
 3. We will not validate entire JSON document in question,
 4. If you are not sure if your JSON is valid, and you'd like to make sure that produced JSON Patch is, use `.validate`.

**Comment by garthk** [comment]
Is it too late to reconsider? 

* The results don't conform to [RFC6902] after being run through `JSON.stringify`
* The results are least surprising to people who never touch `JSON` and most surprising to those that do

[RFC6902]: http://tools.ietf.org/html/rfc6902
[POLA]: http://en.wikipedia.org/wiki/Principle_of_least_astonishment
[Postel]: http://en.wikipedia.org/wiki/Robustness_principle

Like @wolfgang42 in #60, I believe

> it should not be possible for fast-json-patch to emit an invalid json-patch

[RFC6902] says, over and over:

> The operation object [MUST](http://tools.ietf.org/html/rfc2119#section-1) contain a "value" member

If `value` is `undefined` before we `stringify` the patch, we violate the spec. That's not *strictly* our fault: `undefined` literally doesn't exist in [EMCA-404](http://www.ecma-international.org/publications/standards/Ecma-404.htm). I don't think users considering hiring us to help them switch from `POST` to `PATCH` much care for that distinction: they just need the job done. If we can't help them, they'll hire someone else. 

Users can `JSON.stringify` even if they used `undefined`. The results simply lack the `undefined` values and their keys, making `JSON.stringify` [least astonishing] [POLA] and [robust][Postel]. 

As @warpech observed in #60, we can do the same by treating the operation as if the before and after pictures had been run through `JSON.stringify` and `JSON.parse`.

Consider setting a value to `undefined`:

```js
var myobj = { };
var observer = jsonpatch.observe( myobj );
myobj.firstName = undefined;
var patch = jsonpatch.generate( observer );
```

A patch of `[]` might be astonishing to anyone using `jsonpatch` purely in memory, but the whole point of the library is to help users `PATCH` their `JSON`. The `JSON` encodings of `myobj` before and after the mutation are identical. From the `JSON` point of view, the current behaviour is surprising and, once you `stringify` it, non-compliant:

```js
[ { op: 'add', path: '/firstName', value: undefined } ]
```

```json
[ { "op": "add", "path": "/firstName" } ]
```

In short, for any mutation:

* any patch should remain valid after encoding it with `JSON.stringify`
* the patch from `jsonpatch.compare` should be consistent whether you run its inputs through a `stringify` and `parse` round-trip or not
* the patch from `jsonpatch.observe` should match the patch from `jsonpatch.compare`
* all the above should be true even if `undefined` is present

**Comment by sonnyp** [comment]
JSON.stringify discard undefined values in objects, not in arrays where they are turned into null.

```JSON.stringify({foo: undefined})``` returns ```{}```
```JSON.stringify([true, undefined, true])``` returns ```[true,null,true]```

**Comment by warpech** [comment]
Labelling this as "question". This lib will always prioritise performance over sanity-checking. But perhaps something can be done to improve the developer UX.

Currently the explanation in README.md is not very helping (written in ~~strikethrough~~): 

<img width="891" alt="screenshot 2015-10-19 16 39 09" src="https://cloud.githubusercontent.com/assets/566463/10580862/fd0d84d4-767f-11e5-88b3-d0bd53801ed2.png">

**Comment by warpech** [comment]
Fixed in [1.0.0](https://github.com/Starcounter-Jack/JSON-Patch/releases/tag/1.0.0)
